src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1112,6 +1123,148 @@ protected static ImmutableArray<string> CheckTupleElementNames(int cardinality,
+        /// returned details (from the compiler's internal APIs) that are helpful for more precisely diagnosing
+        /// reasons for accessibility failure.</para>
+        /// </remarks>
+        public bool IsSymbolAccessibleWithin(
+            ISymbol symbol,
+            ISymbol within,
+            ITypeSymbol throughType = null)
+        {
+            if (symbol is null)
+            {
+                throw new ArgumentNullException(nameof(symbol));
+            }
+
+            if (within is null)
+            {
+                throw new ArgumentNullException(nameof(within));
+            }
+
+            if (!(within is INamedTypeSymbol || within is IAssemblySymbol))
+            {
+                throw new ArgumentException(CodeAnalysisResources.IsSymbolAccessibleBadWithin, nameof(within));
+            }
+
+            checkInCompilationReferences(symbol, nameof(symbol));
+            checkInCompilationReferences(within, nameof(within));
+            if (!(throughType is null))
+            {
+                checkInCompilationReferences(throughType, nameof(throughType));
+            }
+
+            return IsSymbolAccessibleWithinCore(symbol, within, throughType);
+
+            void checkInCompilationReferences(ISymbol s, string parameterName)
+            {
+                var containingAssembly = computeContainingAssembly(s);
+                if (!assemblyIsInReferences(containingAssembly))
+                {
+                    throw new ArgumentException(string.Format(CodeAnalysisResources.IsSymbolAccessibleWrongAssembly, parameterName), parameterName);
+                }
+            }
+
+            bool assemblyIsInReferences(IAssemblySymbol a)
+            {
+                if (assemblyIsInCompilationReferences(a, this))
+                {
+                    return true;
+                }
+
+                if (this.IsSubmission)
+                {
+                    // Submissions can reference symbols from previous submissions and their referenced assemblies, even
+                    // though those references are missing from this.References. We work around that by digging in
+                    // to find references of previous submissions. See https://github.com/dotnet/roslyn/issues/27356
+                    for (Compilation c = this.PreviousSubmission; c != null; c = c.PreviousSubmission)
+                    {
+                        if (assemblyIsInCompilationReferences(a, c))
+                        {
+                            return true;
+                        }
+                    }
+                }
+
+                return false;
+            }
+
+            bool assemblyIsInCompilationReferences(IAssemblySymbol a, Compilation compilation)
+            {
+                if (a == compilation.Assembly)
+                {
+                    return true;
+                }
+
+                foreach (var reference in compilation.References)
```

Comment:
>foreach (var reference in compilation.References) [](start = 16, length = 49)

I am not comfortable with the approach that iterates through the ```References``` property, this could be error prone and not robust long term. We should simply follow the way compiler works. All assembly symbols that compiler consider as references for the purpose of analysis are in ```this.Assembly.Modules[0].ReferencedAssemblySymbols``` set. It really doesn't matter to the compiler what is in ```References```, that information is used to calculate the  ```this.Assembly.Modules[0].ReferencedAssemblySymbols``` set and it is better if we do not assume how that is done and what else could be added to the set.

Reply:
This implementation directly follows the specification. Your suggestion is a deep violation of the Law of Demeter. I'll think about a different way of specifying this API that would not be such a violation. It is possible that a hypothetical future API on Compilation that exposes the "computed references" would be more apropos, and if such a thing comes to exist this implementation could be changed to use it.

Reply:
> This implementation directly follows the specification. Your suggestion is a deep violation of the Law of Demeter. I'll think about a different way of specifying this API that would not be such a violation. It is possible that a hypothetical future API on Compilation that exposes the "computed references" would be more apropos, and if such a thing comes to exist this implementation could be changed to use it.

The ```this.Assembly.Modules[0].ReferencedAssemblySymbols``` is a public API, I think we should specify the API in these terms instead then. API design meeting didn't say anything about this API depending on ```References``` property. Here is the quote from the notes: "API should require incoming symbols to be “from the compilation” (or a referenced assembly) or throw". I think that ReferencedAssemblySymbols API is a better reflection of " referenced assemblies". If you disagree, let's bring it to the API review meeting and discuss it there. As it stands right now and given my experience in the area the usage of ReferencedAssemblySymbols is more appropriate, less bug prone, more robust to future changes (not likely to require additional maintenance) and is aligned with other places where compilers decide what assemblies are referenced by a compilation.  



---
In reply to: [193180245](https://github.com/dotnet/roslyn/pull/27312#discussion_r193180245) [](ancestors = 193180245)

Reply:
The API review group has confirmed the specification based on `Compilation.References`.

---
In reply to: [193252122](https://github.com/dotnet/roslyn/pull/27312#discussion_r193252122) [](ancestors = 193252122,193180245)